java-checks/src/main/java/org/sonar/java/checks/security/SecureCookieCheck.java

```diff
@@ -19,96 +19,199 @@
-      reportIssue(unsecuredCookie.declaration().simpleName(), "Add the \"secure\" attribute to this cookie");
-    }
+    unsecuredCookies.forEach(v -> reportIssue(v.declaration().simpleName(), MESSAGE));
+    unsecuredReturns.forEach(r -> reportIssue(r, MESSAGE));
+    unsecuredSetters.forEach(m -> reportIssue(m.arguments(), MESSAGE));
   }
 
   @Override
   public void visitNode(Tree tree) {
-    if (hasSemantic()) {
-      if (tree.is(Tree.Kind.VARIABLE)) {
-        VariableTree variableTree = (VariableTree) tree;
-        addToUnsecuredCookies(variableTree);
-      } else if (tree.is(Tree.Kind.METHOD_INVOCATION)) {
-        MethodInvocationTree mit = (MethodInvocationTree) tree;
-        checkSecureCall(mit);
-      }
+    if (!hasSemantic()) {
+      return;
+    }
+    if (tree.is(Tree.Kind.VARIABLE)) {
+      addToUnsecuredCookies((VariableTree) tree);
+    } else if (tree.is(Tree.Kind.ASSIGNMENT)) {
+      addToUnsecuredCookies((AssignmentExpressionTree) tree);
+    } else if (tree.is(Tree.Kind.METHOD_INVOCATION)) {
+      checkSecureCall((MethodInvocationTree) tree);
+    } else {
+      addToUnsecuredReturns((ReturnStatementTree) tree);
     }
   }
 
   private void addToUnsecuredCookies(VariableTree variableTree) {
-    Type type = variableTree.type().symbolType();
-    if (type.is("javax.servlet.http.Cookie") && isConstructorInitialized(variableTree)) {
-      Symbol variableSymbol = variableTree.symbol();
-      //Ignore field variables
-      if (variableSymbol.isVariableSymbol() && variableSymbol.owner().isMethodSymbol()) {
-        unsecuredCookies.add((Symbol.VariableSymbol) variableSymbol);
+    ExpressionTree initializer = variableTree.initializer();
+    // Ignore field variables
+    if (initializer != null && variableTree.symbol().isVariableSymbol() && variableTree.symbol().owner().isMethodSymbol()) {
+      boolean isInitializedWithConstructor = initializer.is(Tree.Kind.NEW_CLASS);
+      boolean isMatchedType = isCookieClass(variableTree.symbol().type()) || isCookieClass(initializer.symbolType());
+      if (isInitializedWithConstructor && isMatchedType && isSecureParamFalse((NewClassTree) initializer)) {
+        unsecuredCookies.add((Symbol.VariableSymbol) variableTree.symbol());
+      }
+    }
+  }
+
+  private void addToUnsecuredCookies(AssignmentExpressionTree assignment) {
+    if (assignment.expression().is(Tree.Kind.NEW_CLASS) && assignment.variable().is(Tree.Kind.IDENTIFIER)) {
+      IdentifierTree identifier = (IdentifierTree) assignment.variable();
+      boolean isMethodVariable = identifier.symbol().isVariableSymbol() && identifier.symbol().owner().isMethodSymbol();
+      boolean isMatchedType = isCookieClass(identifier.symbolType()) || isCookieClass(assignment.expression().symbolType());
+      if (isMethodVariable && isMatchedType && isSecureParamFalse((NewClassTree) assignment.expression())) {
+        unsecuredCookies.add((Symbol.VariableSymbol) ((IdentifierTree) assignment.variable()).symbol());
```

Comment:
This can trigger NPE (even if unlickely...) if the variable is not defined in the same file and visible from this context:

```java
class A { // in A.java
  public UnsecureCookie myCookie;
}

class B extends A { // in B.java

  void foo() {
    myCookie = new UnsecureCookie(...); // will trigger a NPE, the declaration is not available
  }
}